Ensure we do not decode invalid RLPs

jochem-brouwer opened the pull request:
There are some cases where we successfully decode invalid RLPs - we should throw in these cases. In this PR, (rather old) tests from `ethereum/tests` are added, as well some tests from Geth.

This also updates the tests, as they are using deprecated methods like `equal` which should be `strictEqual`.

jochem-brouwer commented:
Note: https://github.com/ethereumjs/rlp/pull/101/commits/f114898110bfae3c3f460239d9a6adc3e51c17ad in this commit I remove one invalid test. If we use Geth's `rlpdump -hex 0x` then it does not throw. Thus an empty buffer is apparently OK as input..?

The majority of the invalid RLP errors come from rather interesting `Buffer.slice(start, end)` behavior. If either `start` or `end` is larger than the length of the Buffer, then this method should not throw. There are some cases where you can produce an RLP which does not throw in our library but are invalid nevertheless.

We also don't have this specific check from [Geth](https://github.com/ethereum/go-ethereum/blob/99be62a9b16fd7b3d1e2e17f1e571d3bef34f122/rlp/decode.go#L945) which I have added as well.

I also had to edit some tests which now throw with a different error message. I'm not super happy with the error messages and we should probably make them more explicit. Am also not entirely convinced that we have now covered all cases, especially because comparing the RLP implementation in Go and in JS is prone to off-by-one errors; in JS we use `Buffer.slice(start, end)` where you thus provide a region where you want to slice, but in Go you have the semantics of `Buffer.slice(start, length)`, i.e. in JS that would be `Buffer.slice(start, start+length)`, and since most times we use the first byte of the RLP to switch on behavior, we thus have to subtract 1 on the reported "length" to actually get the `end` parameter which we should slice into.

jochem-brouwer commented:
Maybe it is an idea to convert this [file](https://github.com/ethereum/go-ethereum/blob/master/rlp/decode_test.go) to JS tests?

ryanio commented:
i think part of the issue with the linting was your master branch was pointed to an older commit.

rebased on master and tidied up `official.spec.ts`

looking good!

ryanio commented:
> Maybe it is an idea to convert this [file](https://github.com/ethereum/go-ethereum/blob/master/rlp/decode_test.go) to JS tests?

that's a good idea, one step further would be to see if any of the cases there are not covered in `ethereum/tests` and have them added upstream for the greater benefit

jochem-brouwer commented:
> > Maybe it is an idea to convert this [file](https://github.com/ethereum/go-ethereum/blob/master/rlp/decode_test.go) to JS tests?
> 
> that's a good idea, one step further would be to see if any of the cases there are not covered in `ethereum/tests` and have them added upstream for the greater benefit

I'll check with ethereum/tests team to see if these RLP tests are managed since the last change was 2+ years ago :) But nevertheless it would be great to see if we have now covered all cases, since fuzzing this thing seems to be rather complicated.

jochem-brouwer commented:
Thanks for the rebase @ryanio :smile: 

I added the test cases from Geth (not all, since some were duplicate or some threw in Go because of type errors, which I have converted).

jochem-brouwer commented:
I'd like at least 2 approvals before merging here, since this is a rather big change.

jochem-brouwer commented:
Two things to do;

- [ ] Check if these new error cases are covered by tests, coverage seems to say no to some cases
- [ ] Cleanup tests, move all invalid tests to `invalid.spec.ts`

jochem-brouwer commented on src/index.ts:
```diff
@@ -166,9 +166,6 @@ function _decode(input: Buffer): Decoded {
     }
     data = safeSlice(input, llength, length + llength)
-    if (data.length < length) {
-      throw new Error('invalid RLP: not enough bytes for string')
```
This can be removed as this exact same check is actually done in `safeSlice`, so `safeSlice` would always throw if `data.length < length` condition is true, and thus this line cannot be reached.

jochem-brouwer commented on src/index.ts:
```diff
@@ -201,9 +198,6 @@ function _decode(input: Buffer): Decoded {
 
     innerRemainder = safeSlice(input, llength, totalLength)
-    if (innerRemainder.length === 0) {
-      throw new Error('invalid rlp, List has a invalid length')
```
This one can also be removed but is a bit more complex. We only have `innerRemainder.length === 0` in case that `llength` and `totalLength` are equal (any other combination would throw `safeSlice`. However, we have `totalLength = llength + length` and therefore we have that `length` should be 0, but a few lines above there is a check whether `length < 56`; if that is the case then it throws. So, this condition can also not be reached.

jochem-brouwer commented:
Coverage increased to more than 95% :) Now it is really ready for review, will not push any changes anymore.

ryanio commented on src/index.ts:
```diff
@@ -177,22 +188,23 @@ function _decode(input: Buffer): Decoded {
-    length = safeParseInt(input.slice(1, llength).toString('hex'), 16)
+    length = safeParseInt(safeSlice(input, 1, llength).toString('hex'), 16)
+    if (length < 56) {
+      throw new Error('invalid rlp: encoded list too short')
```
nit just to match the other error messages
```suggestion
      throw new Error('invalid RLP: encoded list too short')
```

ryanio commented on test/official.spec.ts:
```diff
@@ -1,23 +1,188 @@
+
+function bufferArrayToStringArray(buffer: any): any {
+  return buffer.map((buf: any) => {
+    if (Array.isArray(buf)) {
+      return bufferArrayToStringArray(buf)
+    } else {
+      return buf.toString('hex')
+    }
```
simplification I learned from this rule, although not too big of a deal since this function is just an internal test helper :) https://eslint.org/docs/rules/no-else-return 

```suggestion
    if (Array.isArray(buf)) {
      return bufferArrayToStringArray(buf)
    }
    return buf.toString('hex')
```

ryanio reviewed (approved):
lgtm, nice! great PR. we can wait for one more approval before merging

jochem-brouwer commented:
Thanks for the review and changes, let's indeed wait for one more approval :smile:

acolytec3 commented on src/index.ts:
```diff
@@ -129,7 +143,7 @@ function _decode(input: Buffer): Decoded {
       data = Buffer.from([])
     } else {
-      data = input.slice(1, length)
+      data = safeSlice(input, 1, length)
```
```suggestion
invalid rlp encoding: byte must be less than 0x80
```

Since we're talking about invalid encoding, can we update this language for clarification?

jochem-brouwer replied:
Please see the new error message, this should be more clear :)

acolytec3 replied:
Thanks, much clearer now!

acolytec3 commented on test/dataTypes.spec.ts:
```diff
@@ -72,16 +75,34 @@ describe('RLP encoding (string):', function () {
+      'zoo255zoo255zzzzzzzzzzzzssssssssssssssssssssssssssssssssssssssssssssss',
+    ]
+    const encodedArrayOfStrings = RLP.encode(data)
+    assert.equal(RLP.getLength(encodedArrayOfStrings), 86)
```
```suggestion
assert.strictEqual(RLP.getLength(encodedArrayOfStrings), 86)
```

Since `equal` is deprecated, can we update to the supported equivalent function?  Tests pass fine with this change.

acolytec3 commented on test/official.spec.ts:
```diff
@@ -1,23 +1,187 @@
+            `invalid output: ${gethCase.input}`
+          )
+        } else {
+          assert.equal(output.toString('hex'), gethCase.value, `invalid output: ${gethCase.input}`)
```
```suggestion
          assert.strictEqual(output.toString('hex'), gethCase.value, `invalid output: ${gethCase.input}`)
```

Update to non-deprecated `strictEqual`

acolytec3 reviewed (changes requested):
Left a couple of comments but looks good otherwise!

acolytec3 commented on test/official.spec.ts:
```diff
@@ -1,23 +1,191 @@
+        const output = RLP.decode(buffer)
+        if (Array.isArray(output)) {
+          const arrayOutput = bufferArrayToStringArray(output)
+          assert.equal(
```
```suggestion
 assert.strictEqual(
```

acolytec3 reviewed (approved):
LGTM!
